## SETUP -----------------------------------------------------------------------------------------------------------------------

#Requires R 4.3.2 and Rtools 4.3
#dplyr 1.1.4; httr 1.4.7; BiocManager 1.30.22; rWikiPathways 1.22.1; RCy3 2.22.1
#Cytoscape 3.10.1

setwd("~/GitHub/SCZ-CNV")
  #Setting working directory
rm(list=ls())
  #Cleaning up workspace
packages <- c("dplyr","httr","stringr","gprofiler2","rvest","tidyr")
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
if(!"rWikiPathways" %in% installed.packages()){
  if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
  BiocManager::install("rWikiPathways")
}
if(!"RCy3" %in% installed.packages()){
  if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
  BiocManager::install("RCy3")
}
  #Checking if required packages are installed and installing if not
  #Different structure for rWikiPathways and RCy3 packages as these are not installed directly but via the BiocManager package
invisible(lapply(c(packages,"rWikiPathways","RCy3"), require, character.only = TRUE))
  #Loading libraries

sysdatetime <- Sys.time()
datetime <- format(sysdatetime, format = "%Y-%m-%d_%Hh%M")
dir.create("Outputs")
dir.create(sprintf("Outputs/Session-%s",datetime))
dir.create(sprintf("Outputs/Session-%s/Networks",datetime))
  #Creating directories for outputs generated by this script to be saved in; new "Session" folder created each time the script is ran (contains generated networks, metadata, and sessionInfo)
nw_savepath <- sprintf("%1$s/Outputs/Session-%2$s/Networks/",getwd(),datetime)
file.create(sprintf("Outputs/Session-%s/metadata.txt",datetime))
  #Creating a new metadata file with the current date and time as suffix for easier organisation
  #Such a metadata file should be generated every time this script is ran to record parameters and versions of functions or databases, including the time avoids files being overwritten if the script is run multiple times a day (can even include seconds if script is ran multiple times per minute)
metadata.add <- function(info) {
  write(sapply(info, as.character), sprintf("Outputs/Session-%s/metadata.txt",datetime),append=TRUE, sep = "\n")
}
metadata.add(sysdatetime)
metadata.add(Sys.timezone())
metadata.add("")
  #Adding the timezone, date, and time to the metadata 

invisible(file.create(sprintf("Outputs/Session-%s/sessioninfo-%s.txt",datetime,datetime)))
writeLines(capture.output(sessionInfo()),sprintf("Outputs/Session-%s/sessioninfo-%s.txt",datetime,datetime))
  #Generating and adding a sessionInfo file to the current session output folder

execution_times <- list()

start_section <- function(section_name) {
  execution_times[[section_name]] <<- Sys.time()
}

end_section <- function(section_name) {
  end_time <- Sys.time()
  start_time <- execution_times[[section_name]]
  execution_time <- end_time - start_time
  # Convert elapsed time to seconds
  execution_time_seconds <- as.numeric(execution_time, units = "secs")
  # Convert seconds to a human-readable format with two decimal places
  time_string <- format(round(execution_time_seconds, 2), nsmall = 2)
  # Add "s" for seconds
  time_string <- paste(time_string, "s", sep = "")
  measurements <- paste(section_name, "\t", time_string, "\n", sep = "")
  cat(measurements, file = sprintf("Outputs/Session-%s/execution-times.txt",datetime), append = TRUE)
  # Remove start time from the list
  execution_times[[section_name]] <- NULL
}


cytoscapePing()
cytoscapeVersionInfo()
  #Checking if Cytoscape is running and version info
metadata.add(capture.output(cytoscapeVersionInfo()))

checkinstall.app <- function(app) {
  status_string <- getAppStatus(app)
    #Getting install status of app
  words <- strsplit(status_string, " ")[[1]]
  last_word <- tail(words, 1)
    #getAppStatus returns a character string instead of a logical value, so the last word (usually either "Installed" or "Uninstalled") from the output is checked
  if (last_word == "Installed") {
    print(sprintf("App %s is already installed.",app))
    } else {
    installApp(app)
    print(sprintf("Installed app %s.",app))
    }
}
  #Function to check whether required Cytoscape apps are installed and installing them if not
applist <- c("Wikipathways", "DisGeNET-app", "CyTargetLinker","stringApp","BridgeDb","clusterMaker2")
  #WikiPathways v.3.3.10
  #DisGeNET-app v.7.3.0
  #CyTargetLinker v. 4.1.0
  #stringApp v. 2.0.2
  #BridgeDb v.1.2.0
  #clusterMaker2 v.2.3.4
lapply(applist,checkinstall.app)
  #Checking and installing (if required) necessary Cytoscape apps
lapply(applist,getAppInformation)
metadata.add("Required Cytoscape apps and versions:")
invisible(metadata.add(print(lapply(applist,getAppInformation))))
metadata.add("")

# FUNCTION DICTIONARY-------------------------------------------------------------------------------------------------------------------
.defaultBaseUrl <- 'http://127.0.0.1:1234/v1'
  #Defining the default base URL found in the RCy3 source as R object for altmergeNetworks
altmergeNetworks <- function(               sources = NULL,
                                            title = NULL,
                                            operation = "union",
                                            nodeKeys = NULL,
                                            nodeMergeMap = NULL,
                                            nodesOnly = FALSE,
                                            edgeKeys = NULL,
                                            edgeMergeMap = NULL,
                                            networkMergeMap = NULL,
                                            inNetworkMerge = TRUE,
                                            base.url = .defaultBaseUrl) {
  cmd.string <- 'network merge' # a good start
  
  # sources must be suppled
  if(is.null(sources)) {
    message("Missing sources!")
    return(NULL)
  } else {
    sources.str <- paste(sources, collapse = ",")
    cmd.string <- paste0(cmd.string,' sources="',sources.str,'"')
  }
  
  # defaults
  cmd.string <- paste0(cmd.string,' operation=',operation)
  cmd.string <- paste0(cmd.string,' nodesOnly=',nodesOnly)
  cmd.string <- paste0(cmd.string,' inNetworkMerge=',inNetworkMerge)
  
  # optional args
  if(!is.null(title))
    cmd.string <- paste0(cmd.string,' netName="',title,'"')
  if(!is.null(nodeKeys))
    cmd.string <- paste0(cmd.string,' nodeKeys="',paste(nodeKeys, collapse = ","),'"')
  if(!is.null(edgeKeys))
    cmd.string <- paste0(cmd.string,' edgeKeys="',paste(edgeKeys, collapse = ","),'"')
  if(!is.null(nodeMergeMap)){
    nodeMergeMap.str <- paste(nodeMergeMap, collapse = ",")
    nodeMergeMap.str <- gsub("c\\(", "{", nodeMergeMap.str)
    nodeMergeMap.str <- gsub("\\)", "}", nodeMergeMap.str)
    cmd.string <- paste0(cmd.string,' nodeMergeMap="',nodeMergeMap.str,'"')
  }
  if(!is.null(edgeMergeMap)){
    edgeMergeMap.str <- paste(edgeMergeMap, collapse = ",")
    edgeMergeMap.str <- gsub("c\\(", "{", edgeMergeMap.str)
    edgeMergeMap.str <- gsub("\\)", "}", edgeMergeMap.str)
    cmd.string <- paste0(cmd.string,' edgeMergeMap="',edgeMergeMap.str,'"')
  }
  if(!is.null(networkMergeMap)){
    networkMergeMap.str <- paste(networkMergeMap, collapse = ",")
    networkMergeMap.str <- gsub("c\\(", "{", networkMergeMap.str)
    networkMergeMap.str <- gsub("\\)", "}", networkMergeMap.str)
    cmd.string <- paste0(cmd.string,' networkMergeMap="',networkMergeMap.str,'"')
  }
  
  res.data <- commandsPOST(cmd.string, base.url = base.url)
  
  if(!is.null(res.data$SUID))
    return(res.data$SUID)
  else
    return(res.data)
}
  #Normally, RCy3's 'mergeNetworks' function would be used to unify imported networks into one supernetwork
  #This function does however not work on the latest RCy3 release (v.2.22.1), but does work when running the script on RCy3 v.2.14.2
  #RCy3 2.14.2 requires R v.4.1.3, requiring the entire script to run on an old version of R for one function that is used once
  #Here, we redefine the function using the source code from RCy3 v.2.14.2 and simply use this alternate function to merge networks


sparqlquery <- function(endpoint,queryfile,output) {
  if (tolower(endpoint) %in% c("WikiPathways","wikipathways","wp","WP")) {
  file_path <- paste0(getwd(),sprintf("/Data/WikiPathways/%s",queryfile))
    #Specifying file path to the .txt file containing the query
  base_url <- "https://sparql.wikipathways.org/sparql/?default-graph-uri=&query="
  #Defining the base URL preceding all SPARQL URLs for the endpoint
  } else if (tolower(endpoint) %in% c("AOP-Wiki", "aopwiki","aop-wiki","AOPWiki")) {
    file_path <- paste0(getwd(),sprintf("/Data/AOP-Wiki/%s",queryfile))
    #Specifying file path to the .txt file containing the query
    base_url <- "https://aopwiki.rdf.bigcat-bioinformatics.org/sparql/?default-graph-uri=&query="
    #Defining the base URL preceding all SPARQL URLs for the endpoint
  } else {
    print("Error: Please specify either WikiPathways or AOP-Wiki as endpoint.")
  }
  querybody <- paste(readLines(file_path), collapse = "")
  #Reading a text file containing a SPARQL query stored in the repo
  encoded_query <- URLencode(querybody)
  #Encoding the query as URL
  full_url <- paste0(base_url, encoded_query, "&format=text/html&timeout=0&signal_void=on")
  #Joining the query from the text file and the base URL and adding that output is desired as HTML
  html <- read_html(full_url)
  #Sending the query to the SPARQL endpoint and extracting as HTML
  tablebody <- html %>% 
    html_element("body") %>%
    html_element("div") %>%
    html_element("table")
  #Navigating to the table output by the SPARQL query
  assign(output, html_table(tablebody), envir=.GlobalEnv)
  #Getting the output as tibble
}
  #Function to send a SPARQL query defined in a local text file to the endpoint and extract to desired dataframe

createNodeSource <- function(source,doi=NULL) {
  if (source == "WikiPathways") {
  networkname <- getNetworkName()
  nodetable <- paste0(networkname," default node") 
  } 
    #Networks imported from WikiPathways have a type in the node table designations, as they have 2 spaces between "default" and "node" instead of one
    #This check determines which node table name format is to be applied depending on the source (WikiPathways or other)
  else {
    networkname <- getNetworkName()
    nodetable <- paste0(networkname," default node") 
    }
  commandsRun(sprintf("table create column columnName=fromWikiPathways table=%s type=string",nodetable))
  commandsRun(sprintf("table create column columnName=fromDisGeNET table=%s type=string",nodetable))
  commandsRun(sprintf("table create column columnName=fromPublication table=%s type=string",nodetable))
  commandsRun(sprintf("table create column columnName=Publication.doi table=%s type=string",nodetable))
  commandsRun(sprintf("table create column columnName=fromSTRING table=%s type=string",nodetable))
    #Creating a new column for each source used for all networks
  if ( source == "STRINGnode") {
    commandsRun(sprintf('table set values columnName=%1$s handleEquations=false rowList="selected:true" table=%2$s value=1',source,nodetable))
  }
  else {
  commandsRun(sprintf("table set values columnName=%1$s handleEquations=false rowList=all table=%2$s value=1",source,nodetable))
    #Filling the new column of the corresponding source with 1 to indicate which source the node is imported from 
  }
  if (!is.null(doi)) {
  commandsRun(sprintf("table set values columnName=Publication.doi handleEquations=false rowList=all table=%1$s value=%2$s",nodetable,doi))
    #Adding doi for literature used if provided
  }
}
  #Function to create new column in node table specifying origin of network/node

disgenetRestUrl<-function(netType,host="127.0.0.1",port=1234,version="v7"){
  if(is.null(netType)){
    print("Network type not specified.")
  }else{
    disgeneturl<-sprintf("http://%s:%i/disgenet/%s/%s",host,port,version,netType)
  }
  return (disgeneturl)
}
net <- "gene-disease-net"
disgenetRestUrl(netType = net)
  #Defining object for REST to call DisGeNET automation module; defining that we will be using gene-disease associations (GDA)
disgenetRestCall<-function(netType,netParams){
  disgeneturl<-disgenetRestUrl(netType)
  restCall<-POST(disgeneturl, body = netParams, encode = "json")
  result<-content(restCall,"parsed")
  return(result)
}
  #Object that executes REST calls to DisGeNET module in Cytoscape 
geneDisParams <- function(source,dis,min) {list(
  source = source,
  assocType = "Any",
  diseaseClass = "Any",
  diseaseSearch = dis,
  geneSearch = " ",
  initialScoreValue = min,
  finalScoreValue = "1.0"
)}
  #Specifying parameters of the GDA network to be imported

#queryspecies.wp <- c("Homo sapiens","Rattus norvegicus","Mus musculus")
#getPathways.wp <- function(i) {
#pw <- findPathwaysByText(i)
#pw <- pw %>%
#dplyr::filter(species %in% queryspecies.wp)
#Filtering by species
# pw.ids <- paste0(i, "_wpids")
# assign(pw.ids, as.character(pw$id),envir = .GlobalEnv)
#Extracting WP IDs
#}
#Function to query WikiPathways using keyword and to extract WP IDs for the import function

#import <- function(j) {
#commandsRun(paste0('wikipathways import-as-network id=', j))
#Pasting WikiPathways IDs into a Cytoscape command line prompt to import as networks
#createNodeSource("WikiPathways")
#Filling the 'WikiPathways' column with 1 to indicate the source
#}
#Importing pathways from WikiPathways by pathway ID


# SCHIZOPHRENIA =======================================================================================================================
## IMPORTING AND MERGING ---------------------------------------------------------------------------------------------------------------
start_section("Importing and merging")

sparqlquery("wp","metadataquery.txt","WikiPathways-SPARQL-metadata")
  #Getting the metadata of the endpoint used for the WikiPathways SPARQL queries
metadata.add("WikiPathways SPARQL endpoint metadata")
metadata.add(`WikiPathways-SPARQL-metadata`)
  #Adding the fetched metadata to the metadata file for the session
  #It is technically possible that the metadata would describe an earlier version of the RDF if it is updated while the script runs but this is unlikely

sparqlquery("wp","pathwayquery.txt","wp_pathwaylist")
  #Getting a list of pathways corresponding to a keyword as per defined in the query
writeLines(wp_pathwaylist[["PWID"]], con="Data/WikiPathways/Pathwaylists/automaticpathways.txt")
  #Saving the output pathway list to file
manualpathways <- readLines("Data/WikiPathways/Pathwaylists/manualpathways.txt")
  #Reading a file containing a list of manually selected pathways
allpathways <- c(wp_pathwaylist[["PWID"]],manualpathways)
  #Joining the list of manually and automatically selected pathways together
allpathways_URL <- paste0("<",allpathways,">")
  #Adding <> around all entries for easier use in SPARQL queries
  #URLs need to be surrounded by <> to be recognised as such
writeLines(allpathways_URL, con="Data/WikiPathways/Pathwaylists/allpathways.txt")
  #Writing list of all pathways in SPARQL URL format to file

genedisparams.scz.df <- read.table("Data/DisGeNET/disgenetparams-scz.txt",header=TRUE,sep = "\t")
  #Loading relevant gene-disease networks from DisGeNET
  #Networks of interest manually added into tsv where it is easier to adjust filters
disgeneturl <- c()
  #Preparing container for DisGeNET URL to be saved for addition to metadata file
apply(genedisparams.scz.df,1,function(row) {
  gdp <- geneDisParams(row["source"],row["dis"],row["min"])
  disgeneturl <<- disgenetRestUrl(net)
    #Fetching the DisGeNET URL used to make this call 
  geneDisResult <- disgenetRestCall(net,gdp)
    #Executing the DisGeNET query
  createNodeSource("fromDisGeNET")
    #Adding information about data source to each node
  mapTableColumn("geneId","Human","Entrez Gene","Ensembl")
    #Mapping Entrez Gene IDs to Ensembl IDs
  renameTableColumn("geneName","DisGeNETname")
})  
  #Importing networks from DisGeNET

metadata.add(paste("DisGeNET URL:",disgeneturl))
metadata.add(paste("DisGeNET net type:",net))
metadata.add("")
  #Adding the DisGeNET URL and net type used to add networks to the metadata file


sparqlquery("wp","nodequery.txt","wp_nodelist")
  #Making a SPARQL query to the endpoint to get all nodes associated with a list of pathways
if (any(grepl("identifiers\\.org", wp_nodelist$Identifier))) {
  # Checking whether the 'Identifier' column contains the identifiers.org URL
  #This is to avoid issues later when the identifiers.org part is removed and the code is reran
  wp_nodelist$WPNodeIDType <- gsub(".*/([^/]+)/.*", "\\1", wp_nodelist$Identifier)
  #If 'identifiers.org' is still in the column, extract part of the string into a new column to see what type the identifier is
} else {
  # If 'identifiers.org' is not found, do nothing
}
wp_nodelist[] <- lapply(wp_nodelist, function(x) str_replace_all(x, "https://identifiers\\.org/([^/]+)/", ""))
  #Selecting and removing "https://identifiers.org/xyz" from every row in the df for improved readability
wp_nodelist$CNVassociated <- ifelse(grepl("copy number | CNV",wp_nodelist$PathwayTitle), 1, NA)
  #Adding a new binary column showing if a given node is associated with a CNV based on pathway title
wp_nodelist$WPNodeID <- wp_nodelist$Identifier
  #Generating a duplicate node identifier column since the original column will be lost during Cytoscape import due to it being selected as source column
linebreakrows <- grep("\n",wp_nodelist$name)
  #Some nodes have line breaks in their labels which causes issues in the Cytoscape import
  #The line breaks have been removed, but the problem will persist until the new version of the WikiPathways RDF is released which the SPARQL query accesses
wp_nodelist <- wp_nodelist[-linebreakrows,]
  #Rows with linebreaks are simply removed from the data for now
write.table(wp_nodelist, file=paste0(getwd(),"/Data/WikiPathways/nodelist.tsv"), quote=FALSE, sep="\t", row.names=FALSE)
  #Writing the modified file for Cytoscape import
commandsRun(sprintf('network import file columnTypeList="sa,sa,s,sa,sa,sa,sa,sa" file=%s firstRowAsColumnNames=true rootNetworkList=-- Create new network collection -- startLoadRow=1 delimiters=\\t', paste0(getwd(),"/Data/WikiPathways/nodelist.tsv")))
  #Importing a list of nodes from the output of a WikiPathways SPARQL query (get all nodes in pathways matching the keyword 'Schizophrenia' and some manually selected pathways)
Sys.sleep(0.5)
  #Adding sys.sleep to give Cytoscape sufficient time to import the file as network; otherwise, renaming doesn't always work since no network is selected until the import is complete
renameNetwork("WikiPathways nodes")

sparqlquery("wp","edgequery.txt","wp_edgelist")
  #Making a SPARQL query to the endpoint to get a list of source-target pairs from selected pathways
wp_edgelist[] <- lapply(wp_edgelist, function(x) str_replace_all(x, "https://identifiers\\.org/([^/]+)/", ""))
  #Selecting and removing "https://identifiers.org/xyz" from every row in the df for improved readability
edge_df <- wp_edgelist[grepl("Interaction",wp_edgelist$source) | grepl("Interaction",wp_edgelist$target),]
  #Extracting rows containing "Interaction" in either the source or target column
  #Interaction nodes represent phosphorylation and the like and are not suitable for the network
  #They can still provide information about the connection of gene or other nodes so they can't just be deleted either
  #If an Interaction node is connected to two or more non-interaction nodes, these nodes should be connected to each other, and the interaction node can be deleted
interaction_freq <- table(edge_df$target)
edge_df_filtered <- edge_df[edge_df$target %in% names(interaction_freq[interaction_freq > 1]),]
  #Counting if a certain interaction occurs more than once; this implies that it is connected to more than one non-interaction node
unique_targets <- unique(edge_df_filtered$target)
for (target_val in unique_targets) {
  # Identify rows with duplicate target values
  rows_with_duplicate_target <- which(edge_df_filtered$target == target_val)
  
  if (length(rows_with_duplicate_target) > 1) {
    # Select one of the source values
    source_val_to_transpose <- edge_df_filtered$source[rows_with_duplicate_target[1]]
    
    # Transpose the source value to the target column in the row of the remaining source value
    edge_df_filtered$target[rows_with_duplicate_target[-1]] <- source_val_to_transpose
    
    # Remove duplicate rows
    edge_df_filtered <- edge_df_filtered[-rows_with_duplicate_target[1], ]
  }
}
  #Transposing the non-identifier nodes for source-target pairs; if two nodes are associated with the same interaction, they become source-target pairs
wp_edgelist <- wp_edgelist <- wp_edgelist[!grepl(".*interaction.*", wp_edgelist$source, ignore.case = TRUE) & 
                                            !grepl(".*interaction.*", wp_edgelist$target, ignore.case = TRUE), ]
  #Removing any row containing "Interaction"
wp_edgelist <- rbind(wp_edgelist,edge_df_filtered)
  #Appending the new source-target pairs to the original edge list

write.table(wp_edgelist, file=paste0(getwd(),"/Data/WikiPathways/edgelist.tsv"), quote=FALSE, sep="\t", row.names=FALSE)
  #Writing the modified file for Cytoscape import
commandsRun(sprintf('network import file columnTypeList="sa,s,t" file=%s firstRowAsColumnNames=true rootNetworkList=-- Create new network collection -- startLoadRow=1 delimiters=\\t', paste0(getwd(),"/Data/WikiPathways/edgelist.tsv")))
  #Importing a list of source-target pairs from selected pathways from the ouput of a WikiPathways SPARQL query 
Sys.sleep(0.5)
  #Adding sys.sleep to give Cytoscape sufficient time to import the file as network; otherwise, renaming doesn't always work since no network is selected until the import is complete
renameNetwork("WikiPathways edges")

altmergeNetworks(sources = c("WikiPathways nodes","WikiPathways edges"),
                 title = "WikiPathways networks",
                 operation = "union",
                 nodeKeys=c("WPNodeID","name"))
  #Union merging the node and edge networks to extend the node list with corresponding edges
Sys.sleep(0.5)
createNodeSource("fromWikiPathways")
deleteNetwork('WikiPathways nodes')
deleteNetwork('WikiPathways edges')

Sys.sleep(1)
  #Pausing the script for 1 second - when letting the script run without this, the publication source creation fails

commandsRun(sprintf("network import file columnTypeList='sa,sa,source,sa,sa,sa,sa' file=%s firstRowAsColumnNames=true rootNetworkList=-- Create new network collection -- startLoadRow=1", paste0(getwd(),"/Data/Publications/Trubetskoy.txt")))
  #Importing network from file
  #List of 120 genes implicated in Trubetskoy et al., doi: 10.1038/s41586-022-04434-5
commandsRun("table rename column columnName=Ensembl.ID newColumnName=Ensembl table=Trubetskoy.txt default node")
  #Renaming the Ensembl.ID column from the dataset to Ensembl for coherence with networks from other sources
commandsRun("table rename column columnName=Index.SNP newColumnName=snpID table=Trubetskoy.txt default node")
createNodeSource("fromPublication","10.1038/s41586-022-04434-5")
  #Adding literature as  source to all imported nodes and adding the doi of the corresponding paper
renameNetwork("Trubetskoy risk genes")
  #Renaming the newly imported network
metadata.add("Publications")
metadata.add("Trubetskoy et al. doi: 10.1038/s41586-022-04434-5")
metadata.add("")

# networklist.dup <- getNetworkList()
# dup.filter <- function(input,suffix) {
#   filtered_list <- input[substr(input, nchar(input) - 1,nchar(input))==suffix]}
# duplicates <- dup.filter(networklist.dup,"_1") 
#   #Getting duplicate networks (Cytoscape marks duplicate networks with a "_1" suffix to the network name)
# delete.dupes <- function(nw) {
#  setCurrentNetwork(nw)
#   deleteNetwork()
# }
# lapply(duplicates,delete.dupes)
  #Selecting and deleting duplicate networks

networklist <- getNetworkList()
setCurrentNetwork(networklist[[1]])
for(i in 1:length(networklist)) {
  current <- getNetworkName()
  altmergeNetworks(c(current,networklist[[i]]), paste(current,networklist[[i]]),"union",inNetworkMerge = TRUE,nodeKeys=c("Ensembl","Ensembl"))
}
  #Looping through the network list to merge all currently open networks with each other, creating one large unified network
renameNetwork("Schizophrenia supernetwork")
networklist <- getNetworkList()
snw_scz <- getNetworkName()
  #Getting the name of the unified network to preserve it from deletion
lapply(networklist[networklist != snw_scz],deleteNetwork)
  #Deleting all networks besides newly generated unified network
exportNetwork(filename=paste0(nw_savepath,"SCZ_SNW"),"CX", network = snw_scz, overwriteFile=TRUE)
  #Exporting the supernetwork as cx file

end_section("Importing and merging")


#wpids <- c("4875","5412","4222","4942","5408","5402","5346","5405","5406","5407","4940","4905","5398","5399","4906","4657","4932")
#sczcnv <- sapply(wpids, function(k) paste0("WP",k))
#Manually adding relevant SCZ CNV pathways from WikiPathways

#keyword.wp <- "Schizophrenia"
#getPathways.wp(keyword.wp)
#lapply(c(Schizophrenia_wpids,sczcnv), import)
#Importing WP pathways (both manually added and by keyword). Also adds "WikiPathways" as NodeSource column to node table
#metadata.add(paste("WikiPathways keywords:",keyword.wp))
#metadata.add(paste("WikiPathways manually by ID:",paste(wpids,collapse =", ")))
#metadata.add(paste("WikiPathways queried species:",paste(queryspecies.wp,collapse = ", ")))
#Adding the keyword and species used to filter the WikiPathways query to the metadata file
#metadata.add("")

## FILTERING NETWORK ------------------------------------------------------------------------------------------------------------------
# createColumnFilter(filter.name="type.label",column="Type","Label","IS",apply=FALSE)
# createColumnFilter(filter.name="type.anchor",column="Type","Anchor","IS",apply=FALSE)
# createColumnFilter(filter.name="type.group",column="Type","Group","IS",apply=FALSE)
# createColumnFilter(filter.name="disease.name",column="diseaseName","Schizophrenia","IS")
# createCompositeFilter(filter.name="type.label.anchor.group",c("type.label","type.anchor","type.group","disease.name"),"ANY")
#deleteSelectedNodes()
  #Creating filters and deleting columns in the node table that are not relevant to the supernetwork (leftovers from import sources)
#renameNetwork("SCZ_SNW_filtered")
#snw_scz_filtered <- getNetworkName()
#exportNetwork(filename=paste0(nw_savepath,"SCZ_SNW_filtered"),"CX", network = snw_scz_filtered, overwriteFile=TRUE)
  #Exporting the filtered supernetwork as cx file and tagging it with the time and date made to match with metadata file

## STRING --------------------------------------------------------------------------------------------------------------------------
start_section("STRING")
commandsRun('string stringify colDisplayName=name column=Ensembl compoundQuery=true cutoff=0.9 includeNotMapped=true  networkType="full STRING network" species="Homo sapiens" networkNoGui=current')
# commandsRun('string expand additionalNodes=1000 network=current nodeTypes="Homo sapiens" selectivityAlpha=0.9')
#   #STRINGifying and expanding the network with a 0.9 confidence cutoff (curated information)
# createNodeSource("fromSTRING")
#   #Tagging the newly added nodes as having been sourced from STRING
# mapTableColumn("stringdb::canonical name","Human","Uniprot-TrEMBL","Ensembl",force.single=TRUE)
#   #Mapping stringdb canonical names (Uniprot-TrEMBL identifiers) to Ensembl gene identifiers
#   #This step generates a second Ensembl column ('Ensembl (1)') with ENSG identifiers for the STRING-imported nodes
# renameTableColumn("Ensembl (1)","Ensembldup")
#   #Renaming the duplicate Ensembl column for easier handling
# renameNetwork("SCZ_SNW_filtered_STRING")
# snw_scz_filtered_string <- getNetworkName()
# stringified_nodetable <- paste0(nw_savepath,sprintf("%s node table.csv",snw_scz_filtered_string))
#   #Saving the file path to the node table for easier reading
# commandsRun(sprintf('table export options=CSV outputFile=%1$s table="%2$s default  node"',stringified_nodetable,snw_scz_filtered_string))
#   #Exporting the node table as .csv file to the current session's "Network" folder
# read_stringified_nodetable <- read.csv(stringified_nodetable)
#   #saving the node table as object
# read_stringified_nodetable$Ensembl <- ifelse(read_stringified_nodetable$Ensembl == read_stringified_nodetable$Ensembldup, as.character(read_stringified_nodetable$Ensembl),
#                                ifelse(is.na(read_stringified_nodetable$Ensembl) | read_stringified_nodetable$Ensembl =="",as.character(read_stringified_nodetable$Ensembldup),
#                                       ifelse(is.na(read_stringified_nodetable$Ensembldup) | read_stringified_nodetable$Ensembldup=="",as.character(read_stringified_nodetable$Ensembl),"No Match")))
#   #As the identifier mapping from STRING ENSP to ENSG identifiers generates a second Ensembl column, they are merged into the originial Ensembl column if the contents of the cell match or either is blank
# read_stringified_nodetable = subset(read_stringified_nodetable,select= -Ensembldup)
#   #Removing the duplicate Ensembl column from the table
# write.csv(read_stringified_nodetable, file=stringified_nodetable)
#   #Overwriting the previously exported table with the version containing the merged Ensembl column
# renameTableColumn("@id","X.id")
#   #Renaming the '@id' column in the Cytoscape table to avoid issues when reimporting the .csv (@id is automatically converted to X.id in the CSV)
# loadTableData(read_stringified_nodetable,data.key.column="X.id",table.key.column="X.id")
#   #Reimporting the .csv with the merged Ensembl column to the network as to have ENSG identifiers for almost all nodes 
mapTableColumn("Ensembl","Human","Ensembl","HGNC")
#   #Generating a new column 'HGNC' from Ensembl identifiers - easier and less error-prone than merging various name columns from different import sources
renameTableColumn("HGNC","Name2")
  #Renaming the new 'HGNC' column to 'Name2', which is now to be used as default name column. ('shared name' and 'name' columns are immutable and cannot be deleted or renamed)
marked_cols <- as.list(getTableColumnNames()[!(getTableColumnNames() %in% c("selected","name.copy" ,"SUID","shared name","name","fromDisGeNET","fromWikiPathways","Ensembl","fromPublication","Publication.doi","fromSTRING","CNVassociated","PathwayID","NodeID","NodeIDType","snpID","Name2","DisGeNETname"))])
lapply(marked_cols, function(column) {
  deleteTableColumn(column=column)
})
  #Filtering columns
renameNetwork("SCZ_SNW_STRING")
scz_snw_string <- getNetworkName()
exportNetwork(filename=paste0(nw_savepath,"SCZ_SNW_STRING"),"CX",network=scz_snw_string,overwriteFile=TRUE)
  #Exporting the filtered, stringified supernetwork as cx file and tagging it with the time and data to match with the metadata file
end_section("STRING")

## CLUSTERING ----------------------------------------------------------------------------------------------------------------------
start_section("Clustering")
createColumnFilter(filter.name="delete.noensembl", column="Ensembl","ENSG","DOES_NOT_CONTAIN")
deleteSelectedNodes()
  #Filtering out nodes that do not have an ENSG Ensembl identifier mapped to them
# marked_cols <- as.list(getTableColumnNames()[!(getTableColumnNames() %in% c("selected","name.copy" ,"SUID","shared name","name","Name2","fromDisGeNET","fromWikiPathways","Ensembl","fromPublication","Publication.doi","fromSTRING","CNVassociated"))])
# lapply(marked_cols, function(column) {
#   deleteTableColumn(column=column)
# })
  #Deleting all the columns besides immutable columns, Ensembl, name, and source columns
metadata.add("GLay Clustering")
metadata.add(capture.output(commandsRun('cluster glay clusterAttribute=__glayCluster createGroups=false network=current restoreEdges=true showUI=true undirectedEdges=true')))
  #Clustering the network using the GLay community cluster from the clusterMaker Cytoscape app and recording outcome to metadata
renameNetwork("SCZ_SNW_STRING_clustered")
renameTableColumn('__glayCluster','gLayCluster') 
  #Renaming the newly generated gLayCluster column as the original name with two underscores is not recognized during gene ontology
snw_scz_string_clustered <- getNetworkName()
clustered_nodetable <- paste0(nw_savepath,sprintf("/%s node table.csv",snw_scz_string_clustered))
  #Saving the file path to the node table for easier reading (note the double space between node and table)
commandsRun(sprintf('table export options=CSV outputFile=%1$s table="%2$s default  node"',clustered_nodetable,snw_scz_string_clustered))
  #Exporting the node table as .csv file to the current session's "Network" folder
exportNetwork(filename=paste0(nw_savepath,"SCZ_SNW_STRING_clustered"),"CX",network=snw_scz_string_clustered,overwriteFile=TRUE)
  #Exporting the filtered, stringified, clustered supernetwork as cx file and tagging it with the time and data to match with the metadata file


read_clustered_nodetable <- read.csv(clustered_nodetable)
  #Reading the exported csv
split_df <- split(read_clustered_nodetable$Ensembl,read_clustered_nodetable$gLayCluster)
  #Splitting the node table by cluster
nodecount <- sapply(split_df, length)
  #Counting how many nodes are in each cluster
countmatrix <- matrix(seq(1,length(nodecount)), ncol=1)
countmatrix <- cbind(countmatrix,as.numeric(nodecount))
  #Construcing a matrix showing how many nodes are in each cluster
invalidclusters <- as.list(countmatrix[countmatrix[, 2] < 5, 1])
  #Getting which clusters have fewer than 5 nodes associated with them 
valid_clustered_nodetable <- read_clustered_nodetable[!read_clustered_nodetable$gLayCluster %in% invalidclusters, ]
  #Generating a new df containing only nodes associated with clusters that had 5 or more nodes
split_tbl <- split(valid_clustered_nodetable, valid_clustered_nodetable$gLayCluster)
sourcecount <- function(cluster) {
  wpcount <- sum(split_tbl[[cluster]][["fromWikiPathways"]] == 1, na.rm = TRUE)
  dgcount <- sum(split_tbl[[cluster]][["fromDisGeNET"]] == 1, na.rm = TRUE)
  litcount <- sum(split_tbl[[cluster]][["fromPublication"]] == 1, na.rm = TRUE)
  stringcount <- sum(split_tbl[[cluster]][["fromSTRINGnode"]] == 1, na.rm = TRUE)
  result_df <- data.frame(
    gLayCluster = split_tbl[[cluster]][["gLayCluster"]][1],
    WikiPathways_source = wpcount,
    DisGeNET_source = dgcount, 
    Publication_source = litcount, 
    STRING_source = stringcount
  )
}
sources_count <- do.call(rbind, lapply(seq_along(split_tbl),sourcecount))
  #For each cluster, counting how many nodes are associated with which sources
cnvassociatedcount <- function(cluster) {
  cnvcount <- sum(split_tbl[[cluster]][["CNVassociated"]] == 1,na.rm=TRUE)
  nodecount <- length(split_tbl[[cluster]][["Ensembl"]])
  result_df <- data.frame(
    gLayCluster = split_tbl[[cluster]][["gLayCluster"]][1],
    N_CNVassociated_nodes = cnvcount,
    proportion_CNVassociated_nodes = round((cnvcount / nodecount)*100)
  )
}
cnvassociated_count <- do.call(rbind,lapply(seq_along(split_tbl),cnvassociatedcount))
  #For each cluster, count how many nodes originally come from CNV-associated pathways which pathways they come from 
end_section("Clustering")

## GO ANALYSIS ------------------------------------------------------------------------------------------------------------------------
start_section("GO Analysis")

split_df <- split(valid_clustered_nodetable$Ensembl,valid_clustered_nodetable$gLayCluster)
split_list <- lapply(split_df, as.vector)
  #Splitting the node table by cluster number, i.e. lists of Ensembl IDs are created per cluster
go <- function(cluster) {
  gost(
    query = cluster,
    organism = "hsapiens",
    ordered_query = FALSE,
    multi_query = TRUE,
    significant = TRUE,
    exclude_iea = FALSE,
    measure_underrepresentation = FALSE,
    evcodes = FALSE,
    user_threshold = 0.05,
    correction_method = "g_SCS",
    domain_scope ="annotated",
    custom_bg = NULL,
    numeric_ns = "",
    sources = NULL,
    as_short_link = FALSE,
    highlight = TRUE
  )
}
go_list <- lapply(split_list,go)
  #Iterating the gost GO function over all clusters
saveRDS(go_list, file=paste0(nw_savepath,"/go_list.rds"))
  #Saving the entire generated GO analysis as R object locally
get_top_terms <- function(cluster) {
  terms <- toString(go_list[[cluster]][["result"]][["term_name"]][1:5])
    #Extracting the top 5 term names associated with each cluster
  pval <- toString(go_list[[cluster]][["result"]][["p_values"]][1:5])
    #Extracting the p-values for the corresponding top 5 term names
  nodes <- paste(go_list[[cluster]][["meta"]][["query_metadata"]][["queries"]][["query_1"]],collapse=",")
  nnodes <- str_count(toString(go_list[[cluster]][["meta"]][["query_metadata"]][["queries"]][["query_1"]]),"\\S+")
    #Extracing the number of nodes/genes contained in each cluster
  result_df <- data.frame(
    gLayCluster = cluster, 
    GO_Terms = terms, 
    GO_Pvals = pval,
    Nodes = nodes, 
    N_nodes = nnodes
    )
}
topterms_df <- do.call(rbind, lapply(names(go_list),get_top_terms))
  #Getting top 5 term names and corresponding p-values for each cluster and storing in topterms_df
topterms_df <- cbind(topterms_df,sources_count,cnvassociated_count)
  #joining the cluster table and the table detailing the amount of sources per cluster
write.table(topterms_df, file=paste0(getwd(),"/Data/GO-clusters-vis.tsv"), sep = "\t",row.names=FALSE,quote=FALSE)
  #Writing the table to file for Cytoscape import during visualisation
loadTableData(topterms_df,data.key.column="gLayCluster",table.key.column="gLayCluster")
  #Loading the generated top terms and p-values back to the supernetwork; every gene belonging to cluster x is now associated with the top terms of cluster x
deleteTableColumn('gLayCluster.1')
  #Deleting duplicate gLayCluster column that appears after importing top terms data back to network 
renameNetwork(title=paste0(getNetworkName(),"_GO"))
snw_scz_filtered_string_clustered_go <- getNetworkName()
  #Renaming and saving the network name to indicate addition of GO information

compare_term_id_lists <- function(list1, list2) {
  common_elements <- intersect(list1, list2)
  return(length(common_elements))
}
  #Setting up a function to get intersections between cluster term IDs
match_df <- data.frame(Cluster1 = character(),
                        Cluster2 = character(),
                        Matches = numeric(),
                        stringsAsFactors = FALSE)
  #Setting up a df to store output in 
for (i in 1:(length(go_list) - 1)) {
  for (j in (i + 1):length(go_list)) {
    term_id_i <- go_list[[i]][["result"]][["term_id"]]
    term_id_j <- go_list[[j]][["result"]][["term_id"]]
    
    matches <- compare_term_id_lists(term_id_i, term_id_j)
    
    match_df <- rbind(match_df, data.frame(Cluster1 = names(go_list)[i],
                                             Cluster2 = names(go_list)[j],
                                             Matches = matches))
    #Iterating over go_list to compare GO term IDs between every cluster and store number of overlaps
  }
}
colnames(match_df) <- c("source","target","GO_term_matches")
  #Renaming columns
write.table(match_df, file=paste0(getwd(),"/Data/match_df.tsv"), sep = "\t",row.names=FALSE,quote=FALSE)
exportNetwork(filename=paste0(nw_savepath,"SCZ_SNW_STRING_clustered_GO"),"CX",network=snw_scz_filtered_string_clustered_go,overwriteFile=TRUE)
  #Exporting the filtered, stringified, clustered supernetwork after GO as cx file and tagging it with the time and data to match with the metadata file

end_section("GO Analysis")

##AOP ---------------------------------------------------------------------------------------------------------------------------
start_section("AOP")


importNetworkFromFile(file=paste0(nw_savepath,"SCZ_SNW_STRING_clustered_GO.cx"))
commandsRun('table delete column column="gLayCluster.2" table="SCZ_SNW_STRING_clustered_GO default node')
createColumnFilter(
  filter.name = "has_GO_result",
  column = "N_nodes",
  criterion = 0,
  predicate = "GREATER_THAN",
  anyMatch = TRUE,
  apply = TRUE
)
  #Selecting nodes included in a 'valid' cluster, i.e. clusters with 5 or more nodes (GO analysis only performed for these)
  #N_nodes is only generated for 'valid' clusters, so good column to filter by
invertNodeSelection()
deleteSelectedNodes()
  #Inverting the selection and deleting these nodes: now, the network contains only the nodes that make up the clusters fed into the GO analysis
  #Nodes not associated to a large enough cluster/GO term are likely not involved in any significant SCZ-contributing way 
  #The idea is to link GO terms (formed by clusters/genes) to risk factors, so it wouldn't make sense to also link non-cluster/GO associated nodes


aopprocess <- function(input,tag) {

sparqlquery("AOP-Wiki",input,"keensgpairs")
  #Querying AOP-Wiki for a list of all KEs and associated genes. KEs must be contained in an AOP that has an AO from a list of selected AOs
for (i in 1:ncol(keensgpairs)) {
  for (j in 1:nrow(keensgpairs)) {
    keensgpairs[j, i] <- gsub('"', '', keensgpairs[j, i])
  }
}
  #Removing quotation marks from the df

separate_keensgpairs <- separate_rows(keensgpairs,Ensembl,sep="; ")
  #Dividing comma-separated Ensembl IDs into distinct rows
keensgpairs_byensg <- separate_keensgpairs %>%
  group_by(Ensembl) %>%
  summarise(KEid = paste(KEid, collapse="; "),
            KEtitle = paste(KEtitle, collapse="; "),
            AOid = paste(AOid, collapse="; "),
            AOtitle = paste(AOtitle, collapse = "; "),
            AOPid = paste(AOPid, collapse="; "),
            AOPtitle =paste(AOPtitle, collapse="; "))
  #Concateinating other variables based on unique Ensembl ID to get list of associated KEs, AOs, and AOPs per gene
keensgpairs_byensg_save <- paste0(getwd(),sprintf("/Data/AOP-Wiki/keensgpairs_byensg_%s.tsv",tag))
  #Defining savepath for newly generated df
write.table(keensgpairs_byensg, file=keensgpairs_byensg_save,quote=FALSE, sep="\t", row.names=FALSE)
  #Saving df containing gene-KE-AO-AOP associations to file as tsv for Cytoscape import
commandsRun(sprintf('table import file dataTypeTargetforNetworkCollection="Node Table Columns" delimiters=\\t file=%s firstRowAsColumnNames=true keyColumnForMapping="Ensembl" keyColumnIndex=1 startLoadRow=1',keensgpairs_byensg_save))
  #Importing the gene-KE-AO-AOP table to Cytoscape as table add AOP-Wiki info as node attributes
renameNetwork(paste0(getNetworkName(),"_AOP"))
scz_snw_string_go_aop <- getNetworkName()

commandsRun(sprintf('table export options=CSV outputFile=%s table="SCZ_SNW_STRING_clustered_GO_AOP default  node"',paste0(nw_savepath,"SCZ_SNW_STRING_GO_AOP default node")))
  #Exporting the network table
scz_snw_string_go_aop_node <- read.csv(file=paste0(nw_savepath,"SCZ_SNW_STRING_GO_AOP default node.csv"),header=TRUE)
  #Reading the exported table as Cytoscape object
aop_associated_genes <- scz_snw_string_go_aop_node[!(scz_snw_string_go_aop_node$KEid == ""), , drop=FALSE]
  #Getting which rows (=gene nodes) have info from AOP-Wiki associated to them


objects <- c(keensgpairs,separate_keensgpairs,keensgpairs_byensg,keensgpairs_byensg_save,scz_snw_string_go_aop_node,aop_associated_genes)

object_names <- paste0(deparse(substitute(objects)),'_')

for (i in 1:length(object_names)) {
  assign(paste0(object_names[i],tag),list(objects)[[i]],envir=.GlobalEnv)
}
result_names <- paste(object_names,tag)

}

aopprocess("AO_KE_Ensembl_query.txt","selected")


importNetworkFromFile(file=paste0(nw_savepath,"SCZ_SNW_STRING_clustered_GO.cx"))
commandsRun('table delete column column="gLayCluster.2" table="SCZ_SNW_STRING_clustered_GO default node')
createColumnFilter(
  filter.name = "has_GO_result",
  column = "N_nodes",
  criterion = 0,
  predicate = "GREATER_THAN",
  anyMatch = TRUE,
  apply = TRUE
)
#Selecting nodes included in a 'valid' cluster, i.e. clusters with 5 or more nodes (GO analysis only performed for these)
#N_nodes is only generated for 'valid' clusters, so good column to filter by
invertNodeSelection()
deleteSelectedNodes()
#Inverting the selection and deleting these nodes: now, the network contains only the nodes that make up the clusters fed into the GO analysis
#Nodes not associated to a large enough cluster/GO term are likely not involved in any significant SCZ-contributing way 
#The idea is to link GO terms (formed by clusters/genes) to risk factors, so it wouldn't make sense to also link non-cluster/GO associated nodes

sparqlquery("AOP-Wiki","all_AO_KE_Ensembl_query.txt","all_keensgpairs")
#Querying AOP-Wiki for a list of all KEs and associated genes. KEs must be contained in an AOP that has an AO from a list of selected AOs
for (i in 1:ncol(all_keensgpairs)) {
  for (j in 1:nrow(all_keensgpairs)) {
    all_keensgpairs[j, i] <- gsub('"', '', all_keensgpairs[j, i])
  }
}
#Removing quotation marks from the df

separate_all_keensgpairs <- separate_rows(all_keensgpairs,Ensembl,sep="; ")
#Dividing comma-separated Ensembl IDs into distinct rows
all_keensgpairs_byensg <- separate_all_keensgpairs %>%
  group_by(Ensembl) %>%
  summarise(KEid = paste(KEid, collapse="; "),
            KEtitle = paste(KEtitle, collapse="; "),
            AOid = paste(AOid, collapse="; "),
            AOtitle = paste(AOtitle, collapse = "; "),
            AOPid = paste(AOPid, collapse="; "),
            AOPtitle =paste(AOPtitle, collapse="; "))
#Concateinating other variables based on unique Ensembl ID to get list of associated KEs, AOs, and AOPs per gene
all_keensgpairs_byensg_save <- paste0(getwd(),"/Data/AOP-Wiki/all_keensgpairs_byensg.tsv")
#Defining savepath for newly generated df
write.table(all_keensgpairs_byensg, file=all_keensgpairs_byensg_save,quote=FALSE, sep="\t", row.names=FALSE)
#Saving df containing gene-KE-AO-AOP associations to file as tsv for Cytoscape import
commandsRun(sprintf('table import file dataTypeTargetforNetworkCollection="Node Table Columns" delimiters=\\t file=%s firstRowAsColumnNames=true keyColumnForMapping="Ensembl" keyColumnIndex=1 startLoadRow=1',all_keensgpairs_byensg_save))
#Importing the gene-KE-AO-AOP table to Cytoscape as table add AOP-Wiki info as node attributes
renameNetwork(paste0(getNetworkName(),"_AOP_all"))
scz_snw_string_go_aop <- getNetworkName()


commandsRun(sprintf('table export options=CSV outputFile=%s table="SCZ_SNW_STRING_clustered_GO_AOP_all default  node"',paste0(nw_savepath,"SCZ_SNW_STRING_GO_AOP default node_all")))
#Exporting the network table
scz_snw_string_go_aop_node_all <- read.csv(file=paste0(nw_savepath,"SCZ_SNW_STRING_GO_AOP default node_all.csv"),header=TRUE)
#Reading the exported table as Cytoscape object
aop_associated_genes_all <- scz_snw_string_go_aop_node_all[!(scz_snw_string_go_aop_node_all$KEid == ""), , drop=FALSE]
#Getting which rows (=gene nodes) have info from AOP-Wiki associated to them


summary_go_terms <- read.delim(paste0(getwd(),"/Data/summary_go_terms.txt"),header=TRUE,sep="\t",quote="")
  #Loading cluster titles based on GO terms
aop_associated_genes <- merge(aop_associated_genes,summary_go_terms,"gLayCluster")
  #Assigning cluster titles to right clusters
aop_associated_genes_all <- merge(aop_associated_genes_all, summary_go_terms,"gLayCluster")


separate_aoptitles <- separate_rows(aop_associated_genes,AOPtitle,sep="; ")
aop_freq_table <- table(separate_aoptitles$AOPtitle) 
aop_freq_df <- as.data.frame(aop_freq_table)
add_attributes <- separate_aoptitles %>%
  group_by(AOPtitle) %>%
  summarise (AOPEnsembl = paste(Ensembl,collapse="; "),
             AOPgenename = paste(Name2, collapse="; "),
             AOPsummary_go_term = paste(summary_term.x, collapse="; "))
names(aop_freq_df) <- c("AOPtitle","AOP frequency")
aop_freq_df_full <- merge(aop_freq_df, add_attributes,"AOPtitle")
  #Counting how often which AOPs are associated with all genes

separate_aotitles <- separate_rows(aop_associated_genes,AOtitle,sep="; ")
ao_freq_table <- table(separate_aotitles$AOtitle) 
ao_freq_df <- as.data.frame(ao_freq_table)
add_attributes <- separate_aotitles %>%
  group_by(AOtitle) %>%
  summarise (AOEnsembl = paste(Ensembl,collapse="; "),
             AOgenename = paste(Name2, collapse="; "),
             AOsummary_go_term = paste(summary_term.x, collapse="; "))
names(ao_freq_df) <- c("AOtitle","AO frequency")
ao_freq_df_full <- merge(ao_freq_df, add_attributes,"AOtitle")
#Counting how often which AOPs are associated with all genes

separate_ketitles <- separate_rows(aop_associated_genes,KEtitle,sep="; ")
ke_freq_table <- table(separate_ketitles$KEtitle) 
ke_freq_df <- as.data.frame(ke_freq_table)
add_attributes <- separate_ketitles %>%
  group_by(KEtitle) %>%
  summarise (KEEnsembl = paste(Ensembl,collapse="; "),
             KEgenename = paste(Name2, collapse="; "),
             KEsummary_go_term = paste(summary_term.x, collapse="; "))
names(ke_freq_df) <- c("KEtitle","KE frequency")
ke_freq_df_full <- merge(ke_freq_df, add_attributes,"KEtitle")
#Counting how often which AOPs are associated with all genes

aop_associated_genes_freq <- bind_rows(aop_freq_df_full,ao_freq_df_full,ke_freq_df_full)




separate_aoptitles_all <- separate_rows(aop_associated_genes_all,AOPtitle,sep="; ")
aop_freq_table_all <- table(separate_aoptitles_all$AOPtitle) 
aop_freq_df_all <- as.data.frame(aop_freq_table_all)
add_attributes <- separate_aoptitles_all %>%
  group_by(AOPtitle) %>%
  summarise (AOPEnsembl = paste(Ensembl,collapse="; "),
             AOPgenename = paste(Name2, collapse="; "),
             AOPsummary_go_term = paste(summary_term.x, collapse="; "))
names(aop_freq_df_all) <- c("AOPtitle","AOP frequency")
aop_freq_df_all_full <- merge(aop_freq_df_all, add_attributes,"AOPtitle")
#Counting how often which AOPs are associated with all genes

separate_aotitles_all <- separate_rows(aop_associated_genes_all,AOtitle,sep="; ")
ao_freq_table_all <- table(separate_aotitles_all$AOtitle) 
ao_freq_df_all <- as.data.frame(ao_freq_table_all)
add_attributes <- separate_aotitles_all %>%
  group_by(AOtitle) %>%
  summarise (AOEnsembl = paste(Ensembl,collapse="; "),
             AOgenename = paste(Name2, collapse="; "),
             AOsummary_go_term = paste(summary_term.x, collapse="; "))
names(ao_freq_df_all) <- c("AOtitle","AO frequency")
ao_freq_df_all_full <- merge(ao_freq_df_all, add_attributes,"AOtitle")
#Counting how often which AOPs are associated with all genes

separate_ketitles_all <- separate_rows(aop_associated_genes_all,KEtitle,sep="; ")
ke_freq_table_all <- table(separate_ketitles_all$KEtitle) 
ke_freq_df_all <- as.data.frame(ke_freq_table_all)
add_attributes <- separate_ketitles_all %>%
  group_by(KEtitle) %>%
  summarise (KEEnsembl = paste(Ensembl,collapse="; "),
             KEgenename = paste(Name2, collapse="; "),
             KEsummary_go_term = paste(summary_term.x, collapse="; "))
names(ke_freq_df_all) <- c("KEtitle","KE frequency")
ke_freq_df_all_full <- merge(ke_freq_df_all, add_attributes,"KEtitle")
#Counting how often which AOPs are associated with all genes

aop_associated_genes_freq_all <- bind_rows(aop_freq_df_all_full,ao_freq_df_all_full,ke_freq_df_all_full)



exportNetwork(filename=paste0(nw_savepath,"SCZ_SNW_STRING_GO_AOP"),"CX",network=scz_snw_string_go_aop,overwriteFile=TRUE)
#Renaming and saving the resulting network
end_section("AOP")









# sparqlquery("AOP-Wiki","KERKEquery.txt","KERList")
#   #Sending a query to AOP-Wiki as specified in the query file to get KERs and corresponding KEs for selected AOs
# write.table(KERList, file=paste0(getwd(),"/Data/AOP-Wiki/KERList.tsv"),quote=FALSE,row.names=FALSE,sep="\t")
#   #Writing to file
# 
# sparqlquery("AOP-Wiki","KEtitlequery.txt","KEMap")
#   #Sending a query to AOP-Wiki to get KE and corresponding titles
# write.table(KEMap,file=paste0(getwd(),"/Data/AOP-Wiki/KEMap.tsv"))
#   #Writing to file
# 
# sparqlquery("AOP-Wiki","KEensemblquery.txt","KEEnsembl")
#   #Sending a query to AOP-Wiki to get all KE-Ensembl ID pairings available
# allKEsFromList <- data.frame(allKEs = c(KERList$KEup,KERList$KEdown))
#   #Combining both up- and downregulated KEs into a single column 
# filtered_KEEnsembl <- subset(KEEnsembl, KE %in% allKEsFromList$allKEs)
#   #Removing KE-Ensembl rows that are not found in the selected KEs
# write.table(filtered_KEEnsembl, file=paste0(getwd(),"/Data/AOP-Wiki/KEEnsembl_filtered.tsv"), quote=FALSE,row.names=FALSE,sep="\t" )
#   #Writing the filtered KE-ENSG list to file for Cytoscape loading
# 
# commandsRun(sprintf('network import file columnTypeList="ea,s,t" file=%s firstRowAsColumnNames=true startLoadRow=1 rootNetworkList=-- Create new network collection --',paste0(getwd(),"/Data/AOP-Wiki/KERList.tsv")))
#   #Loading selected KERs into Cytoscape as new network
# commandsRun(sprintf('table import file dataTypeTargetForNetworkCollection="Node Table Columns" delimiters=\\t file=%s firstRowAsColumnNames=true keyColumnForMapping="shared name" keyColumnIndex=1 startLoadRow=1',paste0(getwd(),"/Data/AOP-Wiki/KEMap.tsv")))
#   #Loading a KE mapping file generated from a second SPARQL query that fetches all KE URIs and their titles from AOPwiki into the KER network
#   #This file also had quotation marks removed using a text editor
# commandsRun(sprintf('network import file columnTypeList="s,t" file=%s firstRowAsColumnNames=true startLoadRow=1 rootNetworkList=-- Create new network collection --',paste0(getwd(),"/Data/AOP-Wiki/KEEnsembl_filtered.tsv")))
#   #Loading the filtered KE-ENSG list as new network into Cytoscape
# altmergeNetworks(sources = c('KERList.tsv','KEEnsembl_filtered.tsv'),
#               title='KERs',
#               operation='union'
#               )
#   #Merging the manually curated KERList network and the KE-ENSG list to extend the selected KEs with associated genes
# deleteNetwork(network='KERList.tsv')
# deleteNetwork(network='KEEnsembl_filtered.tsv')
#   #Deleting networks used to make merged network 
# mapTableColumn(
#   column = 'name',
#   species= 'Human',
#   map.from = 'Ensembl',
#   map.to = 'HGNC',
#   force.single = 'true'
# )
# renameTableColumn('HGNC','Name2')
# commandsRun(sprintf('table export options=CSV outputFile=%s table="KERs default node"', paste0(getwd(),"/Data/AOP-Wiki/KE_table.csv")))
#   #Exporting the node table for manipulation via R
# KE_table <- read.csv(paste0(getwd(),"/Data/AOP-Wiki/KE_table.csv"))
#   #Loading the previously exported node table as R object
# KE_table <- KE_table %>%
#   mutate(Ensembl=ifelse(grepl("ENSG",name),name,NA))
#   #Getting and transposing ENSG IDs from the name col to a new 'Ensembl' col to be used as key
# KE_table <- KE_table %>%
#   mutate(KE_URI=ifelse(!grepl("ENSG",name),name,NA))
#   #Getting and transposing KE URIs (all rows not containing 'ENSG') from the name col to a new 'KE_URI' col
#   #These two steps are done as the 'name' and 'shared name' columns will be difficult to work with since they mix both ENSG and other data types
# KE_table <- KE_table %>%
#   mutate(fromAOPwiki=1)
#   #Adding a new column to the node table indicating that the gene nodes in the KE network are imported from AOPwiki
# KE_table <- KE_table %>%
#   select(-shared.name)
#   #Loading the node table in R changes the name the 'shared name' column to 'shared.name' to avoid a space, this results in a duplicate 'shared.name' column when importing the table back to Cytoscape
#   #For this reason, 'shared.name'/'shared name' is simply removed from the df as 'names' can also be used for mapping and contains the same information anyways
# write.csv(KE_table, file=paste0(getwd(),"/Data/AOP-Wiki/KE_table.csv"), row.names=FALSE)
#   #Overwriting the previously exported node table with the updated version
# loadTableData(
#   data = read.table(file=paste0(getwd(),"/Data/AOP-Wiki/KE_table.csv"),header=TRUE, sep =","),
#   data.key.column = "name",
#   table.key.column = "name"
# )
#   #Loading the updated node table back to the network
# Sys.sleep(0.5)
# altmergeNetworks(sources = c('SCZ_SNW_filtered_STRING_clustered_GO','KERs'),
#                  title='SCZ_SNW_filtered_STRING_clustered_GO_KER',
#                  operation='union',
#                  nodeKeys = c('Ensembl','Ensembl')
# )
#   #Merging the KER network with the supernetwork based on Ensembl ID
#   #As this is an union merge, all nodes from the KER network are introduced to the SNW, but the goal was simply to annotate existing genes in the SNW with KEs
#   #Therefore, some filtering is needed
# 
# commandsRun(sprintf('table export options=CSV outputFile=%s table="SCZ_SNW_filtered_STRING_clustered_GO_KER default node"', paste0(getwd(),"/Data/AOP-Wiki/SNW_KER_node.csv")))
# SNW_KER_node <- read.csv(paste0(getwd(),"/Data/AOP-Wiki/SNW_KER_node.csv"))
#   #Saving the SNW node table to file and loading it as R object
# KE_only_nodes <- SNW_KER_node %>%
#   filter(fromAOPwiki == 1 & is.na(fromDisGeNET) & is.na(fromWikiPathways) & is.na(fromPublication) & is.na(fromSTRING)) %>%
#   pull(Ensembl)
#   #Getting a list of nodes that only have AOPwiki as source
#   #This implies that these nodes were not already present in the network since they would've been merged to existing nodes that already had at least one source prior
#   #It is not desired to add new gene nodes to the network from the KER network; gene nodes serve only as references for connecting KEs to already existing genes in the supernetwork
# selectNodes(nodes=KE_only_nodes,
#             by.col="Ensembl")
# deleteSelectedNodes()
#   #Selecting and deleting nodes based on the prior criteria
# 
# commandsRun(sprintf('table export options=CSV outputFile=%s table="SCZ_SNW_filtered_STRING_clustered_GO_KER default edge"', paste0(getwd(),"/Data/AOP-Wiki/SNW_KER_edge.csv")))
# SNW_KER_edge <- read.csv(paste0(getwd(),"/Data/AOP-Wiki/SNW_KER_edge.csv"))
#   #Exporting the SNW edge table to file and loading it as R object
# valid_KE_ensg <- SNW_KER_edge[grepl("ENSG", SNW_KER_edge$name) & grepl("aop.events", SNW_KER_edge$name),]
#   #Getting a list of KEs that do have interactions with genes
# valid_KE <- str_extract(valid_KE_ensg$name, "https://identifiers.org/aop.events/\\S+")
# valid_KE <- unique(valid_KE)
#   #Getting the URIs of the KEs that have interactions with genes
# bad_KE_df <- SNW_KER_edge %>%
#   filter(
#     grepl("aop\\.events", name, ignore.case=TRUE) &
#       !grepl(paste(valid_KE, collapse="|"), name)
#   )
#   #Getting a list of the remaining KEs that are not associated with a gene or with a KE that is associated with a gene (to preserve KERs)
# bad_KE_list <- str_extract_all(bad_KE_df$name, "https://identifiers\\.org/aop\\.events/\\d+")
# bad_KE <- unique(unlist(bad_KE_list))
#   #Getting a list of unique bad KEs
# selectNodes(nodes=bad_KE, by.col="name")
# deleteSelectedNodes()
# 
# sparqlquery("AOP-Wiki","AOPAOquery.txt","AOPs_AOs_for_selected_KEs")
# write.table(AOPs_AOs_for_selected_KEs,file=paste0(getwd(),"/Data/AOP-Wiki/AOPs_AOs_for_selected_KEs.tsv"))
# commandsRun(sprintf('network import file columnTypeList="s,ta,t,x,x" file=%s firstRowAsColumnNames=true startLoadRow=1 rootNetworkList="-- Create new network collection --" delimiters=\\t', paste0(getwd(),"/Data/AOP-Wiki/AOPs_AOs_for_selected_KEs.tsv")))
#   #Importing a list of AOPs associated with selected KEs that have prior been imported to the SNW  
#   #The network contains numerous duplicate edges resulting from the way the SPARQL query is formulated
# Sys.sleep(1)
# commandsRun('analyzer remove duplicated edges createColumn=false ignoreDirection=false network=current')
#   #Removing duplicate edges from current network
#   #RCy3 has a deleteDuplicateEdges function but it seems bugged since it always just removes 1 duplicate edge at a time
#   #Using the command line function removes all duplicate edges right away and avoids having to run a filter to select duplicate edges which is much slower
# renameNetwork("KE-AOP")
# commandsRun(sprintf('network import file columnTypeList="x,sa,s,ta,t" file=%s firstRowAsColumnNames=true startLoadRow=1 rootNetworkList="-- Create new network collection --" delimiters=\\t',paste0(getwd(), "/Data/AOP-Wiki/AOPs_AOs_for_selected_KEs.tsv")))
# commandsRun('analyzer remove duplicated edges createColumn=false ignoreDirection=false network=current')
#   #Removing duplicate edges from current network
# renameNetwork("AOP-AO")
# altmergeNetworks(sources=c("KE-AOP","AOP-AO"),
#                  title='KE-AOP-AO',
#                  operation='union',
#                  nodeKeys=c('name','name'))
#   #Merging the AO and AOP networks 
#   #Necessary to first import as separate networks and then merge as only one source and target col can be selected during network import
# altmergeNetworks(sources=c('KE-AOP-AO','SCZ_SNW_filtered_STRING_clustered_GO_KER'),
#                  title='SCZ_SNW_filtered_STRING_clustered_GO_AOP',
#                  operation='union',
#                  nodeKeys=c('name','name'))
#   #Merging the AOP-AO network to the SNW-KE network to extend KEs connected to risk genes with corresponding AOPs/AOs
# deleteNetwork(network='KE-AOP-AO')
# deleteNetwork(network='KERs')
# deleteNetwork(network='SCZ_SNW_filtered_STRING_clustered_GO_KER')
# deleteNetwork(network='KE-AOP')
# deleteNetwork(network='AOP-AO')
#   #Deleting networks that are no longer needed, used to generate SNW-AOP network
# SNW_AOP <- getNetworkName()
# exportNetwork(filename=paste0(nw_savepath,SNW_AOP),"CX",network=SNW_AOP)
#   #Exporting the SNW containing the AOP extension

##RAW SNW VISUALISATION --------------------------------------------------------------------------------------------------------------------------- 
setCurrentNetwork(snw_scz_filtered_string_clustered_go)
createColumnFilter(
  filter.name = "has_GO_result",
  column = "N_nodes",
  criterion = 0,
  predicate = "GREATER_THAN",
  anyMatch = TRUE,
  apply = TRUE
)
  #Selecting nodes included in a 'valid' cluster, i.e. clusters with 5 or more nodes (GO analysis only performed for these)
  #N_nodes is only generated for 'valid' clusters, so good column to filter by
invertNodeSelection()
deleteSelectedNodes()
  #Inverting the selection and deleting these nodes: now, the network contains only the nodes that make up the clusters fed into the GO analysis
  #Nodes not associated to a large enough cluster/GO term are likely not involved in any significant SCZ-contributing way 
  #The idea is to link GO terms (formed by clusters/genes) 


#These changes to the visualization are essentially only visible when Cytoscape is set to always render details
#The network is big and not really organized besides clusters, the pie chart visualization that follows gives a much better impression

createVisualStyle("SNW_vis")
  #Creating a new visual style that is subsequently customized
setVisualStyle("SNW_vis")
  #Applying the new visual style
clusters <- as.character(unique(getTableColumns("node","gLayCluster"))$gLayCluster)
  #gLayCluster column has many repeats of the clusters but a list of unique clusters is needed for table.column.values
clustercolors <- as.list(paletteColorRandom(length(clusters)))
  #Generating a set of random colors based on the number of clusters in the SNW
  #Make this constant, generate 100 colors or so once 
  #Then just select first N where N=Nclusters, so it stays flexible when more/fewer clusters are generated
  #Much easier for legend too
setNodeColorMapping(table.column = "gLayCluster",
                    table.column.values=clusters,
                    colors=clustercolors,
                    mapping.type="d",
                    default.color="#FF5555",
                    style.name = "SNW_vis"
                    )
  #table.column values must be defined like this as the gLayCluster column has many repeats of cluster numbers
  #Random colors are generated per cluster based on the number of different clusters in the SNW
setEdgeLineWidthDefault(new.width = 0.5,
                        style.name = "SNW_vis")
  #Reducing edge width to decrease hairball effect
setNodeLabelMapping(table.column = "Name2",
                    style.name="SNW_vis")
  #Changing node labels to show HGNC name

##PIE CHART VISUALISATION -----------------------------------------------------------------------------------------------------------------------
commandsRun(sprintf("network import file columnTypeList='s,sa,sa,sa,sa,sa,sa,sa,sa,sa,sa,sa,sa,sa' file=%s firstRowAsColumnNames=true delimiters=\\t rootNetworkList=-- Create new network collection -- startLoadRow=1", paste0(getwd(),"/Data/GO-clusters-vis.tsv")))
#Importing the previoulsy generated table 'GO-clusters-vis' back to Cytoscape as new network
#Essential to use .tsv and importing as such to avoid conflicts generated by .csv - commas separating terms in a string are interpreted as different columns by Cytoscape
Sys.sleep(1)
renameNetwork("GO_Visualisation_SCZ_SNW")
loadTableData(
  data = read.table(file=paste0(getwd(),"/Data/summary_go_terms.txt"),header=TRUE, sep ="\t"),
  data.key.column = "gLayCluster",
  table.key.column = "gLayCluster"
)
#Loading an additional column into the network containing manually created summaries of GO terms based on biological knowledge
commandsRun(sprintf("network import file columnTypeList='s,t,ea' file=%s firstRowAsColumnNames=true delimiters=\\t startLoadRow=1", paste0(getwd(),"/Data/match_df.tsv")))
#Loading edge information (i.e. intersections between GO terms for each cluster)
createVisualStyle(style.name = 'GO_vis')
#Creating a new visual style that is subsequently customized
setVisualStyle("GO_vis")
#Applying a basic visual style to the network 
setNodeSizeMapping(
  table.column = "N_nodes",
  sizes = c(50,200),
  mapping.type = "c",
  style.name= "GO_vis"
)
#Setting the node size proportional to the number of nodes making up a cluster, e.g. a cluster containing more nodes is bigger
setNodeCustomPieChart(
  columns = c("DisGeNET_source","Publication_source","WikiPathways_source","STRING_source"),
  colors = paletteColorBrewerPaired(value.count = 8),
  style.name= "GO_vis"
)
#Turning nodes into pie charts showing which sources make up the proportions of the nodes in them
setNodeShapeDefault("ellipse",
                    style.name= "GO_vis")
setNodeFillOpacityDefault(
  new.opacity = 0,
  style.name= "GO_vis"
)
#Removing node background
setNodeBorderWidthDefault(
  new.width=10,
  style.name= "GO_vis"
)
setNodeBorderColorMapping(
  table.column = "proportion_CNVassociated_nodes",
  table.column.values = c(0,100),
  colors = c("#FFFFFF","#FF0000"),
  mapping.type = "c",
  default.color = "#000000",
  style.name = "GO_vis"
)
setNodeLabelMapping(
  table.column = "summary_term",
  style.name= "GO_vis"
)
#Changing node label to the processes/terms the nodes represent
setNodeFontSizeDefault(
  new.size = 20,
  style.name= "GO_vis"
)
#Setting node font size
setNodeLabelPositionDefault(
  new.nodeAnchor = "S",
  new.graphicAnchor = "N",
  new.justification = "c",
  new.xOffset = "0",
  new.yOffset = "0",
  style.name= "GO_vis"
)
#Moving label
setEdgeLineWidthMapping(
  table.column = "GO_term_matches",
  table.column.values = c(0,861),
  widths = c(0,50),
  mapping.type = "c",
  style.name= "GO_vis"
)
#Setting edge width proportional to number of shared GO terms 
setEdgeOpacityDefault(
  new.opacity = 70,
  style.name= "GO_vis"
)
#Decreasing edge opacity
setEdgeColorDefault('#DD3497',
                    style.name= "GO_vis")
#Changing edge color
commandsRun('layout force-directed defaultEdgeWeight=0.5 defaultNodeMass=3 defaultSpringCoefficient=1e-4 edgeAttribute="GO_term_matches" defaultSpringLength=300 isDeterministic=true maxWeightCutoff=1.79769E308 minWeightCutoff=0E0 numIterations=200 type=Heuristic')
layoutNetwork('force-directed')  
#Adding network layout
#scaleLayout(axis="Both Axes", scaleFactor = 0.95)
#Not working as requires Cytoscape v.3.10.2 which does not seem available yet?
renameNetwork("Supernetwork functional analysis")
deleteNetwork(network="GO_Visualisation_SCZ_SNW")
go_vis_nw <- getNetworkName()
exportNetwork(filename=paste0(nw_savepath,"GO_Visualisation_SCZ_SNW"),"CX",network=go_vis_nw,overwriteFile=TRUE)
fitContent()
exportImage(filename = paste0(getwd(),"/Visualisations/SNW_functional_analysis"),type="SVG", overwriteFile=TRUE, zoom="200")
#Exporting the visualisation as network and as svg



